ships/0025-event-driven-builds.md

```diff
@@ -0,0 +1,321 @@
+    * Expects the commit ID and branch name in the `params`.
+    * Based on the above information, the controller would generate the following:
+          * Read the `Build` in the namespace matching the `Run` CR's `.spec.ref`
+          * A `BuildRun` referencing the above `Build`.
+     
+      Sample `Run` CR the controller would be reacting to:
+       ```
+       
+         - apiVersion: tekton.dev/v1alpha1
+           kind: Run
+           metadata:
+             generateName: build-execution-
+           spec:
+             ref:
+               apiVersion: shipwright.io/v1alpha1
+               kind: Build
+               name: build-cr-name
+        ```
+
+
+#### Webhook endpoint creation process
+
+ 
+Upon creation of a `Build` resource by the user, the following would occur:
+
+1. The Shipwright Build Controller would create a `TriggerTemplate` that would map the information coming in from the event into the "parameters" 
+   expected by the `Run` resource ie, the branch name & the commit ID.
+   
+ ```
+ 
+### Generated by the `Build` reconciler. 
+ 
+apiVersion: triggers.tekton.dev/v1alpha1
+kind: TriggerTemplate
+metadata:
+ name: build-cr-name
+spec:
+ params:
+   - name: branch
+   - name: commit
+   - name: url
+ resourceTemplates:
+   - apiVersion: tekton.dev/v1alpha1
+     kind: Run
+     metadata:
+       generateName: build-execution-
+     spec:
+       ref:
+         apiVersion: shipwright.io/v1alpha1
+         kind: Build
+         name: build-cr-name
+       timeout: 3000s
+       params:
+         - name: revision
+           value: $(tt.params.branch)
+         - name: commit
+           value: $(tt.params.commit)
+         - name: url
+           value: $(tt.params.url)
+ ```
+
+2. The Shipwright Build Controller would create a `EventListener` under-the-hood per build. 
+    
+```
+### Generated by the `Build` reconciler. 
+
+apiVersion: triggers.tekton.dev/v1alpha1
+kind: EventListener
+metadata:
+ name: build-cr-name
+spec:
+ serviceAccountName: pipeline
+ triggers:
+   - bindings:
+       - ref: github-shipwright-webhook
+     template:
+       name: build-cr-name
+   
+```
+
+And that's it, you may now go ahead and 
+
+### Test Plan
+
+To be filled.
+
+### Release Criteria
+
+To be filled.
+
+#### Removing a deprecated feature [if necessary]
+
+N/A
+
+#### Upgrade Strategy [if necessary]
+
+N/A
+
+### Risks and Mitigations
+
+1. This feature opens up the possiblility of triggering Build executions for branches which weren't explicitly specified. 
+
+The custom Tekton controller is responsible for ignoring any requests which would have originated from branches not explicitly 
+specified in the `Build` CR.
+
+
+2. Exposing a webhook URL enables creation of pods ( ie, processes on the node ) by actors who may not necessarily have access to the cluster.
+
+This does open up an attack vector given the execution of the build is done using the configured service account. 
```

Comment:
Given we are dealing with external access, we need to be clear that it does require special attention from the cluster admin.